File: src/main/java/arriba/fix/disruptor/SessionNotifyingFixMessageEntryBatchHandler.java
package arriba.fix.disruptor;

import arriba.fix.messages.FixMessage;
import arriba.fix.session.Session;
import arriba.fix.session.SessionResolver;
import com.lmax.disruptor.EventHandler;

public final class SessionNotifyingFixMessageEntryBatchHandler implements EventHandler<FixMessageEntry> {

    private final SessionResolver sessionResolver;

    public SessionNotifyingFixMessageEntryBatchHandler(final SessionResolver sessionResolver) {
        this.sessionResolver = sessionResolver;
    }

  @Override
  public void onEvent(FixMessageEntry entry, boolean b) throws Exception {
        final FixMessage fixMessage = entry.getFixMessage();

        final Session session = this.sessionResolver.resolve(fixMessage.getSessionId());

        session.onMessage(fixMessage);
    }

    public void onEndOfBatch() throws Exception {}

}

File: src/main/scala/arriba/scala/fix/disruptor/SessionNotifyingFixMessageEntryBatchHandler.scala
package arriba.scala.fix.disruptor

import arriba.scala.fix.session.{SessionResolver, Session}
import com.lmax.disruptor.BatchHandler
import arriba.scala.fix.messages.FixMessage

final class SessionNotifyingFixMessageEntryBatchHandler(sessionResolver: SessionResolver) extends BatchHandler[FixMessageEntry] {

  def onAvailable(entry: FixMessageEntry) {
    val fixMessage: FixMessage = entry.getFixMessage
    val session: Session = this.sessionResolver.resolve(fixMessage.getSessionId)
    session.onMessage(fixMessage)
  }

  def onEndOfBatch() {}
}

